src/middlewares/check_auth.rs

```diff
@@ -0,0 +1,303 @@
+    auth_routes: Vec<String>,
+}
+
+impl CheckAuthToken {
+    pub fn new(auth_routes: Vec<String>) -> Self {
+        CheckAuthToken { auth_routes }
+    }
+}
+
+impl<S: 'static, B> Transform<S, ServiceRequest> for CheckAuthToken
+where
+    S: Service<ServiceRequest, Response = ServiceResponse<B>, Error = Error> + 'static,
+    S::Future: 'static,
+    B: 'static,
+{
+    type Response = ServiceResponse<EitherBody<B>>;
+    type Error = Error;
+    type InitError = ();
+    type Transform = CheckAuthTokenMiddleware<S>;
+    type Future = Ready<Result<Self::Transform, Self::InitError>>;
+
+    fn new_transform(&self, service: S) -> Self::Future {
+        ready(Ok(CheckAuthTokenMiddleware {
+            service: Rc::new(service),
+            auth_routes: self.auth_routes.clone(),
+        }))
+    }
+}
+pub struct CheckAuthTokenMiddleware<S> {
+    service: Rc<S>,
+    auth_routes: Vec<String>,
+}
+
+const AUTH_TOKEN_HEADER: &str = "authorization";
+
+fn is_auth_route(routes: &[String], path: &str) -> bool {
+    routes.iter().any(|x| *x == path)
+}
+
+#[derive(Debug)]
+pub struct UserId(String);
+
+impl<S: 'static, B> Service<ServiceRequest> for CheckAuthTokenMiddleware<S>
+where
+    S: Service<ServiceRequest, Response = ServiceResponse<B>, Error = Error>,
+    S::Future: 'static,
+    B: 'static,
+{
+    type Response = ServiceResponse<EitherBody<B>>;
+    type Error = Error;
+    type Future = LocalBoxFuture<'static, Result<Self::Response, Self::Error>>;
+
+    dev::forward_ready!(service);
+
+    fn call(&self, request: ServiceRequest) -> Self::Future {
+        if !is_auth_route(&self.auth_routes, request.path()) {
+            let res = self.service.call(request);
+            return Box::pin(async { res.await.map(ServiceResponse::map_into_left_body) });
+        }
+
+        let token = if let Some(header) = request.headers().get(AUTH_TOKEN_HEADER) {
+            match header.to_str() {
+                Ok(header) => {
+                    let split_header_bearer = header.split(' ').collect::<Vec<&str>>();
+                    let token = split_header_bearer.get(1);
+                    if let Some(token) = token {
+                        token.to_owned()
+                    } else {
+                        ""
+                    }
+                }
+                Err(_) => "",
+            }
+        } else {
+            ""
+        };
+
+        if token.is_empty() {
+            let (request, _pl) = request.into_parts();
+
+            let response = HttpResponse::BadRequest().finish().map_into_right_body();
+
+            return Box::pin(async { Ok(ServiceResponse::new(request, response)) });
+        }
+
+        let token = token.to_string();
+
+        let components = request.app_data::<Data<AppComponents>>().unwrap().clone();
+        let svc = self.service.clone();
+
+        Box::pin(async move {
+            let user_id = {
+                let mut user_cache = components.users_cache.lock().unwrap();
```

Comment:
There's a clippy warning here about the MutexGuard through awaits and suggesting to use an async-aware Mutex instead. Do we know if this affects us?